tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,518 @@
+    tf.compat.v1.placeholder(dtype=output_tensor.dtype,
+                             shape=output_tensor.shape,
+                             name=node.name)
+    return sess.graph_def.node[0]
+
+
+def _get_input_names(subgraph: graph_pb2.GraphDef) -> List[Text]:
+  input_names = [node.name for node in subgraph.node
+                 if _is_placeholder_op(node)]
+  return input_names
+
+
+def _get_non_input_names(subgraph: graph_pb2.GraphDef) -> List[Text]:
+  non_input_names = [node.name for node in subgraph.node
+                     if not _is_placeholder_op(node)]
+  return non_input_names
+
+
+def _partition_one_remote_op_layer(
+    remote_op_layer: List[Text],
+    node_name_to_node_def: Mapping[Text, node_def_pb2.NodeDef]
+) -> List[ExecutionSpec]:
+  """Constructs ExecutionSpecs for a remote op layer.
+
+  As discussed in _get_execution_specs(), a remote op layer contains one
+  or more remote ops having no dependencies on each other. However, instead of
+  having one ExecutionSpec to store a layer (as it is with subgraph layer),
+  we use multiple ExecutionSpecs to represent a remote op layer.
+
+  Args:
+    remote_op_layer: A list of remote op names for a remote op layer.
+    node_name_to_node_def: A mapping from node names to `NodeDef` protos.
+
+  Returns:
+    A list of ExecutionSpecs representing a remote op layer.
+  """
+  list_of_specs = []
+
+  for remote_op_name in remote_op_layer:
+    spec = ExecutionSpec(
+        subgraph=None,
+        input_names=list(node_name_to_node_def[remote_op_name].input),
+        output_names=[remote_op_name],
+        is_remote_op=True)
+    list_of_specs.append(spec)
+
+  return list_of_specs
+
+
+def _modify_execution_specs(specs: List[ExecutionSpec]) -> None:
+  """Modifies the execution specs.
+
+  For an execution spec (representing a partitioned subgraph), a placeholder
+  input may either be a remote op, a placeholder input of the original graph,
+  or a regular node from the previously visited subgraphs. In the last
+  scenario, we need to check if that regular node is one of the output nodes
+  of a previously visited subgraph. If not, we'll add it to the output nodes.
+
+  Args:
+    specs: A list of ExecutionSpecs, where each ExecutionSpec represents a
+           partitioned subgraph, and the order of the list represents the
+           order of the execution.
+  """
+  for current_spec_index, current_spec in enumerate(specs):
+    for previous_spec in specs[:current_spec_index]:
+      if previous_spec.is_remote_op:
+        continue
+      _modify_output_names(current_spec, previous_spec)
+
+
+def _modify_output_names(current_spec: ExecutionSpec,
+                         previous_spec: ExecutionSpec) -> None:
+  for input_name in current_spec.input_names:
+    if input_name in _get_non_input_names(previous_spec.subgraph):
+      if input_name in previous_spec.output_names:
+        continue
+      previous_spec.output_names.append(input_name)
```

Comment:
This algorithm is incorrect because the input that this layer requires could be an output from a few layers ago.

Example 1:
```
a*  b*
 \ /
  c
```
The specs are `[{a*}, {b*}, {c}]`. Note that the third spec has `a*` as an input, which happens two specs ago.

Example 2:
```
  a
 /  \
b*  c
 \ /
  d
```
The specs are `[{a}, {b*}, {c, d}]`. Note that the third spec has `a` as an input, which happens two specs ago.

Reply:
To clarify, `*` indicate remote op nodes.

Reply:
As we discussed yesterday, outputs from previous (computed) layers within the same graph are saved in Beam elements. So the inputs of a current layer should always be ready.